vault/backend.go

```diff
@@ -0,0 +1,616 @@
+// Copyright (C) 2015-2022 Lightning Labs and The Lightning Network Developers
+// Copyright (C) 2022 Bottlepay and The Lightning Network Developers
+
+package vault
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/ecdsa"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcutil/hdkeychain"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func (b *backend) listAccounts(ctx context.Context, req *logical.Request,
+	data *framework.FieldData) (*logical.Response, error) {
+
+	strNode := data.Get("node").(string)
+
+	seed, net, err := b.getNode(ctx, req.Storage, strNode)
+	if err != nil {
+		b.Logger().Error("Failed to retrieve node info",
+			"node", strNode, "error", err)
+		return nil, err
+	}
+	defer zero(seed)
+
+	rootKey, err := hdkeychain.NewMaster(seed, net)
+	if err != nil {
+		return nil, err
+	}
+	defer rootKey.Zero()
+
+	acctList := "{\n    \"accounts\": [\n"
+
+	listAccount := func(purpose, coin, act uint32, addrType string,
+		version []byte) (string, error) {
+
+		strListing := ""
+
+		// Derive purpose.
+		purposeKey, err := rootKey.DeriveNonStandard(
+			purpose + hdkeychain.HardenedKeyStart,
+		)
+		if err != nil {
+			return "", err
+		}
+		defer purposeKey.Zero()
+
+		// Derive coin.
+		coinKey, err := purposeKey.DeriveNonStandard(
+			coin + hdkeychain.HardenedKeyStart,
+		)
+		if err != nil {
+			return "", err
+		}
+		defer coinKey.Zero()
+
+		// Derive account.
+		actKey, err := coinKey.DeriveNonStandard(
+			act + hdkeychain.HardenedKeyStart,
+		)
+		if err != nil {
+			return "", err
+		}
+		defer actKey.Zero()
+
+		// Get account watch-only pubkey.
+		xPub, err := actKey.Neuter()
+		if err != nil {
+			return "", err
+		}
+
+		// Ensure we get the right HDVersion for the account key.
+		if version != nil {
+			xPub, err = xPub.CloneWithVersion(version)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		strPurpose := fmt.Sprintf("%d", purpose)
+		strCoin := fmt.Sprintf("%d", coin)
+		strAct := fmt.Sprintf("%d", act)
+
+		strListing += "        {\n"
+
+		strListing += "            \"name\": \""
+		if act == 0 {
+			strListing += "default"
+		} else {
+			strListing += "act:" + strAct
+		}
+		strListing += "\",\n"
+
+		strListing += "            \"address_type\": \"" + addrType +
+			"\",\n"
+
+		strListing += "            \"extended_public_key\": \"" +
+			xPub.String() + "\",\n"
+
+		strListing += "            \"master_key_fingerprint\": null,\n"
+
+		strListing += "            \"derivation_path\": \"m/" +
+			strPurpose + "'/" + strCoin + "'/" + strAct + "'\",\n"
+
+		strListing += "            \"external_key_count\": 0,\n"
+
+		strListing += "            \"internal_key_count\": 0,\n"
+
+		strListing += "            \"watch_only\": false\n"
+
+		strListing += "        }"
```

Comment:
This and other calls look like they can benefit from defining annotated go types and returning those to vault
?

Reply:
Fixed. :heavy_check_mark:

I also updated the JSON sample format in README.md, it's more compact (but not as pretty).

Reply:
Isn't this independent of the change to using annotated go types, because previously vault also returned json?

Reply:
Yes, but I didn't indent/pretty-print the JSON. If it matters, I can do that again.